### PR TITLE
Process watchdog

### PR DESCRIPTION
**Pull request by lgolding**
@jaredpar @tannergooding @jasonmalinowski 

Tracks a process and all its descendants. If the process and all descendants don't complete in the specified time, take a screenshot, get crash dumps from all remaining processes, and terminate them.

Sample usage:

`ProcessWatchdog --time-limit 5 --executable notepad.exe --output-folder C:\Users\lgolding\Desktop\ProcessWatchdogOutput --procdump-path "F:\Program Files\Procdump\procdump.exe" --screenshot`

or with short switches:

`ProcessWatchdog -t 5 -e notepad.exe -o C:\Users\lgolding\Desktop\ProcessWatchdogOutput -d "F:\Program Files\Procdump\procdump.exe" -s`

**NOTE** The code as it stands has a known race condition. When I periodically determine the descendants of the root process, I use WMI to query for processes whose `ParentProcessId` property matches the root process (I do this recursively). But by the time I make a `Process` object from the process id returned from WMI, that process might have terminated and another one with the same id might have been created. Tanner has a suggestion for this but I want to get this PR out there while I work on it.

**UPDATE** The race condition is now fixed by use of a new `UniqueProcess` class.

### CODE REVIEW

File: Roslyn.sln [hunk 1]
```diff
@@ -352,28 +354,28 @@ Global
 		src\Compilers\Core\CommandLine\CommandLine.projitems*{ad6f474e-e6d4-4217-91f3-b7af1be31ccc}*SharedItemsImports = 13
 		src\Compilers\VisualBasic\BasicAnalyzerDriver\BasicAnalyzerDriver.projitems*{e8f0baa5-7327-43d1-9a51-644e81ae55f1}*SharedItemsImports = 13
 		src\Compilers\Core\AnalyzerDriver\AnalyzerDriver.projitems*{d0bc9be7-24f6-40ca-8dc6-fcb93bd44b34}*SharedItemsImports = 13
-		src\Compilers\Core\CommandLine\CommandLine.projitems*{06b26dcb-7a12-48ef-ae50-708593abd05f}*SharedItemsImports = 4
 		src\Compilers\Server\ServerShared\ServerShared.projitems*{06b26dcb-7a12-48ef-ae50-708593abd05f}*SharedItemsImports = 4
+		src\Compilers\Core\CommandLine\CommandLine.projitems*{06b26dcb-7a12-48ef-ae50-708593abd05f}*SharedItemsImports = 4
```
Comment (lgolding) [inline Roslyn.sln]: I didn't change any of this, not sure what's going on here. I just added the ProcessWatchdog project... oh, it seems to have just moved this line down by one line. ok.

File: src/Tools/ProcessWatchdog/Options.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using CommandLine;
```
Comment (lgolding) [inline src/Tools/ProcessWatchdog/Options.cs]: Using the CommandLineParser NuGet package that Michael and I have used in several other projects. Saves all that pesky hand-parsing.

File: src/Tools/ProcessWatchdog/ProcessTracker.cs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+            _trackedProcesses = _trackedProcesses.Where(tp => !tp.HasExited).ToList();
+
+            // CAUTION: This code is subject to a race condition where between one
+            // call to update and the next, all the processes in the list ended, but new
+            // processes (which we are not yet tracking) were created. In that case,
+            // _trackedProcesses would now be empty, and the ProcessWatchdog would exit,
+            // even though there are still processes we care about.
+            //
+            // This should not happen for the scenarios we care about, since the parent
+            // process should outlive all its descendants.
```
Comment (lgolding) [inline src/Tools/ProcessWatchdog/ProcessTracker.cs]: This comment does not refer to the race condition I mentioned in the PR description. This is another, more benign race condition that (as I said in the comment here) I think doesn't actually occur in the scenario we care about.

File: src/Tools/ProcessWatchdog/ProcessTracker.cs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                int[] existingProcessIds = _trackedProcesses.Select(tp => tp.Process.Id).ToArray();
+
+                foreach (Process descendant in GetDescendants(_parentProcess.Id))
+                {
+                    // BUG: This code is subject to a race condition where, between the time we
+                    // captured the existing process ids and the time we enumerated the descendants,
+                    // one of those existing processes had terminated and a new descendant process
+                    // was created with the same id. This can happen, since process ids can be
+                    // recycled.
+                    if (!existingProcessIds.Contains(descendant.Id))
```
Comment (lgolding) [inline src/Tools/ProcessWatchdog/ProcessTracker.cs]: This is the race condition I mentioned in the PR description.

File: src/Tools/ProcessWatchdog/ScreenShotSaver.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+using System;
+using System.ComponentModel;
+using System.Drawing;
+using System.Drawing.Imaging;
+using System.IO;
+using System.Windows.Forms;
+
+namespace ProcessWatchdog
+
```
Comment (jaredpar) [inline src/Tools/ProcessWatchdog/ScreenShotSaver.cs]: Extra blank line.

Comment (lgolding) [inline src/Tools/ProcessWatchdog/ScreenShotSaver.cs]: Fixed.

**Comment by lgolding** [comment]
Ping @jaredpar @tannergooding @jasonmalinowski 

(I know Tanner is otherwise occupied right now...)

**Comment by jasonmalinowski** [comment]
@dotnet/roslyn-ide would also probably have good thoughts here.

File: src/Tools/ProcessWatchdog/ProcessTracker.cs [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            foreach (TrackedProcess trackedProcess in _trackedProcesses)
+            {
+                // Launch another procdump process, distinct from the one that has been
+                // monitoring the tracked process. This procdump process will take an
+                // immediate dump of the tracked process.
+                Process immediateDumpProcess = _procDump.DumpProcessNow(
+                    trackedProcess.Process.Id,
+                    trackedProcess.Description);
+
+                while (!immediateDumpProcess.HasExited)
```
Comment (tannergooding) [inline src/Tools/ProcessWatchdog/ProcessTracker.cs]: `immediateDumpProcess.WaitForExit()`?

**Comment by tannergooding** [comment]
I gave this is a rough overview and things look good. I'll try and give it a finer grained lookover later if I get the time.

File: src/Tools/ProcessWatchdog/ProcessTracker.cs (hunk 6 above)
Comment (lgolding) [inline src/Tools/ProcessWatchdog/ProcessTracker.cs]: Better, thanks, fixed.

**Pull request merged**